### PR TITLE
fixes #22555; implements `newStringUninit`

### LINKED ISSUE #22555: `newString` zeroes memory contrary to its documentation, significantly slowing it down
### Description

`newString` is supposed to return an uninitialized string, and thus presumably be fast:

https://github.com/nim-lang/Nim/blob/d677ed31e50a322851b476f20fd1719eb17cd426/lib/system.nim#L976

However, the actual implementations zero the full string leading to significant overhead making the function unfit for its purpose:

https://github.com/nim-lang/Nim/blob/devel/lib/system/strs_v2.nim#L113

### Nim Version

1.6, devel, both refc and orc

### Current Output

_No response_

### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #22555; alternative to https://github.com/nim-lang/Nim/pull/22562

close https://github.com/nim-lang/Nim/pull/22562

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -1630,6 +1610,40 @@ when notJSnotNims and defined(nimSeqsV2):
...
+    ##   x[0] = 10
+    ##   ```
+    result = newSeqOfCap[T](len)
+    when defined(nimSeqsV2):
+      cast[ptr int](addr result)[] = len
+    else:
+      var s = cast[PGenericSeq](result)
+      s.len = len
+
+  func newStringUninitialized*(len: Natural): string =
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
  proc newStringUninitialized*(len: Natural): string =
```

File: lib/system.nim [hunk 2]
```diff
@@ -1630,6 +1610,40 @@ when notJSnotNims and defined(nimSeqsV2):
...
+    ## Only available for numbers types. Note that the sequence will be
+    ## uninitialized. After the creation of the sequence you should assign
+    ## entries to the sequence instead of adding them.
+    ## Example:
+    ##   ```nim
+    ##   var x = newSeqUninitialized[int](3)
+    ##   assert len(x) == 3
+    ##   x[0] = 10
+    ##   ```
+    result = newSeqOfCap[T](len)
```
Comment (ringabout) [inline lib/system.nim:1626]: I have made changes, which ensure that `newSeqOfCap` doesn't initialize every slot => https://github.com/nim-lang/Nim/pull/22568

File: lib/system.nim [hunk 3]
```diff
@@ -1630,6 +1610,40 @@ when notJSnotNims and defined(nimSeqsV2):
...
+      s.len = len
+
+  proc newStringUninitialized*(len: Natural): string =
+    ## Returns a new string of length `len` but with uninitialized
+    ## content. One needs to fill the string character after character
+    ## with the index operator `s[i]`.
+    ##
+    ## This procedure exists only for optimization purposes;
+    ## the same effect can be achieved with the `&` operator or with `add`.
+    result = newStringOfCap(len)
```
Comment (ringabout) [inline lib/system.nim:1640]: P.S. I have made changes, which ensure that `newStringOfCap` doesn't initialize every slot => https://github.com/nim-lang/Nim/pull/22568

**Comment by Araq** [comment]
Is the suffix `Uninitialized` already in use? Because if not I prefer the shorter `Uninit`.

**Comment by ringabout** [comment]
Okay, I just kept the same suffix with `newSeqUninitialized`.

File: changelog.md [hunk 4]
```diff
@@ -11,6 +11,8 @@
 
 [//]: # "Additions:"
 
+- Adds `newStringUninitialized` to system, which creates a new string of length `len` like `newString` but with uninitialized content.
```
Comment (ringabout) [inline changelog.md]:
```suggestion
- Adds `newStringUninit` to system, which creates a new string of length `len` like `newString` but with uninitialized content.
```

File: lib/system.nim [hunk 5]
```diff
@@ -1630,6 +1610,40 @@ when notJSnotNims and defined(nimSeqsV2):
...
+    ##   x[0] = 10
+    ##   ```
+    result = newSeqOfCap[T](len)
+    when defined(nimSeqsV2):
+      cast[ptr int](addr result)[] = len
+    else:
+      var s = cast[PGenericSeq](result)
+      s.len = len
+
+  proc newStringUninitialized*(len: Natural): string =
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
  proc newStringUninit*(len: Natural): string =
```

File: tests/system/tnewstring_uninitialized.nim [hunk 6]
```diff
@@ -0,0 +1,11 @@
+discard """
+  matrix: "--mm:refc;"
+"""
+
+# bug #22555
+var x = newStringUninitialized(10)
```
Comment (ringabout) [inline tests/system/tnewstring_uninitialized.nim]:
```suggestion
var x = newStringUninit(10)
```

**Comment by ringabout** [comment]
> Is the suffix Uninitialized already in use?

Well, there is a pre-existing function `newSeqUninitialized`. I thought it might be good to keep the same suffix.

**Comment by Araq** [comment]
`newSeqUninitialized` sucks too and should be renamed too (with deprecation process yada yada).

**Comment by ringabout** [comment]
Absolutely, that will be changed after my next PR.

**Pull request merged**